src/beignet/splits/_murcko_scaffold_split.py

```diff
@@ -0,0 +1,110 @@
+
+try:
+    from rdkit import Chem
+    from rdkit.Chem.Scaffolds.MurckoScaffold import MurckoScaffoldSmiles
+
+    _RDKit_AVAILABLE = True
+except (ImportError, ModuleNotFoundError):
+    _RDKit_AVAILABLE = False
+    Chem, MurckoScaffoldSmiles = None, None
+
+
+def murcko_scaffold_split(
+    dataset: Dataset,
+    smiles: list[str],
+    test_size: float | int,
+    *,
+    seed: int = 0xDEADBEEF,
+    shuffle: bool = True,
+    include_chirality: bool = False,
+) -> tuple[Subset, Subset]:
+    """Split a dataset based on Murcko scaffold splitting based
+    on provided SMILES strings.
+
+    Note that for datasets that are small or not highly diverse,
+    the final test set may be smaller than the specified test_size.
+
+    Parameters:
+        dataset (Dataset): The dataset to split.
+        smiles (list[str]): A list of SMILES strings.
+        test_size (float | int): The size of the test set.
+            If float, should be between 0.0 and 1.0.
+            If int, should be between 0 and len(smiles).
+        seed (int): The random seed to use for shuffling.
+        shuffle (bool): Whether to shuffle the indices.
+        include_chirality (bool): Whether to include chirality in the scaffold.
+
+    Returns:
+        tuple[Subset, Subset]: The train and test subsets.
+
+    References:
+        - Bemis, G. W., & Murcko, M. A. (1996).
+        The properties of known drugs. 1. Molecular frameworks.
+        Journal of medicinal chemistry, 39(15), 2887–2893.
+        https://doi.org/10.1021/jm9602928
+        - "RDKit: Open-source cheminformatics. https://www.rdkit.org"
+
+    """
+    train_idx, test_idx = _murcko_scaffold_split_indices(
+        smiles,
+        test_size,
+        seed=seed,
+        shuffle=shuffle,
+        include_chirality=include_chirality,
+    )
+    return Subset(dataset, train_idx), Subset(dataset, test_idx)
+
+
+def _murcko_scaffold_split_indices(
+    smiles: list[str],
+    test_size: float | int,
+    *,
+    seed: int = 0xDEADBEEF,
+    shuffle: bool = True,
+    include_chirality: bool = False,
+) -> tuple[list[int], list[int]]:
+    """Get train and test indices based on Murcko scaffold splitting."""
+    if not _RDKit_AVAILABLE:
+        raise ImportError(
+            "This function requires RDKit to be installed (pip install rdkit)"
+        )
+
+    if (
+        isinstance(test_size, int) and (test_size <= 0 or test_size >= len(smiles))
+    ) or (isinstance(test_size, float) and (test_size <= 0 or test_size >= 1)):
+        raise ValueError(
+            f"Test_size should be a float in (0, 1) or and int < {len(smiles)}."
+        )
+
+    if isinstance(test_size, float):
+        test_size = math.ceil(len(smiles) * test_size)
+
+    scaffolds = defaultdict(list)
+
+    for ind, s in enumerate(smiles):
+        mol = Chem.MolFromSmiles(s)
+        if mol is not None:
+            scaffold = MurckoScaffoldSmiles(mol=mol, includeChirality=include_chirality)
```

Comment:
Weird function? It looks like a wrapper around:

```
Chem.MolToSmiles(GetScaffoldForMol(mol), includeChirality)
```

No clue about this:

https://github.com/rdkit/rdkit/blob/3457c1eb60846ea821e4a319f3505933027d3cf8/rdkit/Chem/Scaffolds/MurckoScaffold.py#L77-L85

It looks autogenerated?

Reply:
yeah `GetScaffoldForMol` looks strange